cache/filecache/filecache.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package filecache provides an implementation of prototransform.Cache
+// that is based on the file system. Cached data is stored in and loaded
+// from files, with cache keys being used to form the file names.
+//
+// This is the simplest form of caching when sharing cache results is
+// not needed and the workload has a persistent volume (e.g. if the
+// application restarts, it will still have access to the same cache
+// files written before it restarted).
+package filecache
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/bufbuild/prototransform"
+)
+
+type Config struct {
+	// Required: the folder in which cached files live.
+	Path string
+	// Defaults to "cache_" if left empty. This is added to the
+	// cache key and the extension below to form a file name.
+	// A trailing underscore is not necessary and will be added
+	// if not present (to separate prefix from the rest of the
+	// cache key).
+	FilenamePrefix string
+	// Defaults to ".bin" if left empty. This is added to the
+	// cache key and prefix above to form a file name.
+	FilenameExtension string
+	// The mode to use when creating new files in the cache
+	// directory. Defaults to 0600 if left zero. If not left
+	// as default, the mode must have at least bits 0400 and
+	// 0200 (read and write permissions for owner) set.
+	FileMode fs.FileMode
+}
+
+// New creates a new cache that stores items in files. The given
+// path must be a directory, in which the files will be stored.
+func New(config Config) (prototransform.Cache, error) {
```

Comment:
Why the constructor? Why not just call it `Cache` and make the default value usable? Returning concrete values is more common in std lib and OSS for go

Reply:
> Returning concrete values is more common in std lib and OSS for go

Sure, but factory methods are also common in std lib and OSS, for types that have invariants or where zero value is not obviously useful.

In this case, not knowing where to put the cache makes the zero value less useful. And the default you suggested I think is more surprising than intuitive (and likely not even appropriate in production environments). So the factory method allows us to validate the invariants that a valid path has been provided and that it is writable. That way, their application likely would fail to start if the had a bad cache configuration. The other way (just let them use struct literal to configure, with no extra validation) means that their application could think it has successfully started up but actually be broken and fail every attempt to write to the cache.